File: jussi/validators.py
# -*- coding: utf-8 -*-
import itertools as it
from typing import NoReturn

import structlog

from jussi.request.jsonrpc import JSONRPCRequest

#from .errors import InvalidRequest
from .errors import JussiCustomJsonOpLengthError
from .errors import JussiLimitsError
from .typedefs import JrpcRequest
from .typedefs import JrpcResponse
from .typedefs import RawRequest
from .typedefs import SingleJrpcResponse

logger = structlog.get_logger(__name__)

GET_BLOCK_RESULT_KEYS = {"previous",
                         "timestamp",
                         "witness",
                         "transaction_merkle_root",
                         "extensions",
                         "witness_signature",
                         "transactions",
                         "block_id",
                         "signing_key",
                         "transaction_ids"}

JSONRPC_REQUEST_KEYS = {'id', 'jsonrpc', 'method', 'params'}
JSONRPC_RESPONSE_KEYS = {'id', 'jsonrpc', 'result', 'error'}

NONE_TYPE = type(None)
ID_TYPES = (int, str, float, NONE_TYPE)
PARAMS_TYPES = (list, dict, NONE_TYPE)

CUSTOM_JSON_SIZE_LIMIT = 8192
CUSTOM_JSON_FOLLOW_RATE = 2

BROADCAST_TRANSACTION_METHODS = {
    'broadcast_transaction',
    'broadcast_transaction_synchronous'
}


#
# validate_* methods raise on invalid input
#


def validate_jsonrpc_request(request: RawRequest) -> NoReturn:
    from .errors import InvalidRequest
    if isinstance(request, dict):
        assert JSONRPC_REQUEST_KEYS.issuperset(request.keys()) and \
            request['jsonrpc'] == '2.0' and \
            isinstance(request['method'], str) and \
            isinstance(request.get('id', None), ID_TYPES) and \
            isinstance(request.get('params', None), (list, dict, type(None)))
    elif isinstance(request, list) and request:
        for r in request:
            assert JSONRPC_REQUEST_KEYS.issuperset(r.keys()) and \
                r['jsonrpc'] == '2.0' and \
                isinstance(r['method'], str) and \
                isinstance(r.get('id'), ID_TYPES) and \
                isinstance(r.get('params'), PARAMS_TYPES)
    elif isinstance(request, JSONRPCRequest):
        pass  # already be validated
    else:
        raise InvalidRequest(request=request)


#
# is_valid_* methods return True or False, but they don't raise
#


def is_valid_single_jsonrpc_response(response: SingleJrpcResponse) -> bool:
    # response.get('jsonrpc')  == '2.0' and \
    return isinstance(response, dict) and \
        (bool('result' in response) ^ bool('error' in response)) and \
        {'id', 'jsonrpc', 'result', 'error'}.issuperset(response.keys())


def is_valid_non_error_single_jsonrpc_response(response: SingleJrpcResponse) -> bool:
    # response.get('jsonrpc') == '2.0' and
    return isinstance(response, dict) and \
        'result' in response and \
        {'id', 'jsonrpc', 'result'}.issuperset(response.keys())


def is_valid_non_error_jussi_response(
        request: JrpcRequest,
        response: JrpcResponse) -> bool:
    try:
        if isinstance(request, JSONRPCRequest):
            if not is_valid_non_error_single_jsonrpc_response(response):
                return False
            if is_get_block_request(request):
                return is_valid_get_block_response(request, response)
            return True
        if isinstance(request, list):
            return len(response) > 0 and \
                isinstance(response, list) and \
                len(request) == len(response) and \
                all(is_valid_non_error_jussi_response(req, resp)
                    for req, resp in zip(request, response))
        return False
    except Exception as e:
        logger.error('is_valid_non_error_jussi_response error', e=e,
                     jid=request.jussi_request_id)
    return False


def is_get_block_request(request: JSONRPCRequest) -> bool:
    return request.urn.namespace in ('steemd', 'appbase') and request.urn.method == 'get_block'


def is_get_block_header_request(request: JSONRPCRequest) -> bool:
    return request.urn.namespace in (
        'steemd', 'appbase') and request.urn.method == 'get_block_header'


def is_get_dynamic_global_properties_request(request: JSONRPCRequest) -> bool:
    return request.urn.namespace in (
        'steemd', 'appbase') and request.urn.method == 'get_dynamic_global_properties'


def is_valid_get_block_response(
        request: JSONRPCRequest,
        response: SingleJrpcResponse) -> bool:
    if not is_get_block_request(request) and \
            is_valid_non_error_single_jsonrpc_response(response):
        return False
    request_block_num, response_block_num = 'nope', 'nope'
    try:
        params = request.urn.params
        if isinstance(params, list):
            request_block_num = params[0]
        elif isinstance(params, dict):
            request_block_num = params['block_num']
        else:
            raise ValueError(f'bad urn params from {request}: {params} ')

        if 'result' not in response:
            raise Exception('jsonrpc response did not contain result')
        elif response['result'] is None:
            return False  # block does not exist yet

        if 'block_id' in response['result']:
            block_id = response['result']['block_id']
        elif 'block' in response['result']:
            block_id = response['result']['block']['block_id']
        else:
            return False
        response_block_num = block_num_from_id(block_id)
        assert int(request_block_num) == response_block_num
        return True
    except KeyError as e:
        logger.error('is_valid_get_block_response key error',
                     e=e,
                     jid=request.jussi_request_id,
                     response=response)
    except AssertionError:
        logger.error('request_block != response block_num',
                     jid=request.jussi_request_id,
                     request_block_num=request_block_num,
                     response_block_num=response_block_num)
    except Exception as e:
        logger.error('is_valid_get_block_response error', e=e,
                     jid=request.jussi_request_id, )
    return False


def is_broadcast_transaction_request(request: JSONRPCRequest) -> bool:
    return request.urn.method in BROADCAST_TRANSACTION_METHODS


def limit_broadcast_transaction_request(request: JSONRPCRequest, limits=None) -> NoReturn:
    if is_broadcast_transaction_request(request):
        if isinstance(request.urn.params, list):
            request_params = request.urn.params[0]
        elif isinstance(request.urn.params, dict):
            request_params = request.urn.params['trx']
        else:
            raise ValueError(
                f'Unknown request params type: {type(request.urn.params)} urn:{request.urn}')
        ops = [op for op in request_params['operations'] if op[0] == 'custom_json']
        if not ops:
            return
        blacklist_accounts = set()
        try:
            blacklist_accounts = limits['accounts_blacklist']
        except Exception as e:
            logger.warning('using empty accounts_blacklist', e=e,
                           jid=request.jussi_request_id, )

        limit_custom_json_op_length(ops, size_limit=CUSTOM_JSON_SIZE_LIMIT)
        limit_custom_json_account(ops, blacklist_accounts=blacklist_accounts)


def limit_custom_json_op_length(ops: list, size_limit=None):
    if any(len(op[1]['json'].encode('utf-8')) > size_limit for op in ops):
        raise JussiCustomJsonOpLengthError(size_limit=size_limit)


def limit_custom_json_account(ops: list, blacklist_accounts=None):
    accts = set(
        it.chain.from_iterable(op[1]["required_posting_auths"] for op in ops))
    if not accts.isdisjoint(blacklist_accounts):
        raise JussiLimitsError()


def block_num_from_id(block_hash: str) -> int:
    """return the first 4 bytes (8 hex digits) of the block ID (the block_num)
    """
    return int(str(block_hash)[:8], base=16)


def jsonrpc_cache_key(request: JSONRPCRequest) -> str:
    return str(request.urn)

File: internal/validators/validators.go
package validators

import (
	"fmt"
	"reflect"
	"strconv"

	"github.com/steemit/jussi/internal/errors"
	"github.com/steemit/jussi/internal/request"
)

// JSONRPCRequestKeys are the valid keys in a JSON-RPC request
var JSONRPCRequestKeys = map[string]bool{
	"id":      true,
	"jsonrpc": true,
	"method":  true,
	"params":  true,
}

// JSONRPCResponseKeys are the valid keys in a JSON-RPC response
var JSONRPCResponseKeys = map[string]bool{
	"id":      true,
	"jsonrpc": true,
	"result":  true,
	"error":   true,
}

// IDTypes are valid types for JSON-RPC ID field
var IDTypes = []reflect.Kind{
	reflect.String,
	reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
	reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64,
	reflect.Float32, reflect.Float64,
}

// ParamsTypes are valid types for JSON-RPC params field
var ParamsTypes = []reflect.Kind{
	reflect.Slice,
	reflect.Map,
	reflect.Invalid, // nil
}

// ValidateJSONRPCRequest validates a JSON-RPC 2.0 request
func ValidateJSONRPCRequest(request interface{}) error {
	switch req := request.(type) {
	case map[string]interface{}:
		return validateSingleRequest(req)
	case []interface{}:
		if len(req) == 0 {
			return fmt.Errorf("batch request cannot be empty")
		}
		for i, r := range req {
			if err := validateSingleRequest(r.(map[string]interface{})); err != nil {
				return fmt.Errorf("batch request[%d]: %w", i, err)
			}
		}
		return nil
	default:
		return fmt.Errorf("request must be an object or array")
	}
}

func validateSingleRequest(req map[string]interface{}) error {
	// Check required fields
	jsonrpc, ok := req["jsonrpc"].(string)
	if !ok || jsonrpc != "2.0" {
		return fmt.Errorf("jsonrpc must be '2.0'")
	}

	method, ok := req["method"].(string)
	if !ok {
		return fmt.Errorf("method must be a string")
	}
	if method == "" {
		return fmt.Errorf("method cannot be empty")
	}

	// Check id (optional but must be valid type if present)
	if id, ok := req["id"]; ok && id != nil {
		if !isValidIDType(id) {
			return fmt.Errorf("id must be string, number, or null")
		}
	}

	// Check params (optional but must be valid type if present)
	if params, ok := req["params"]; ok && params != nil {
		if !isValidParamsType(params) {
			return fmt.Errorf("params must be array, object, or null")
		}
	}

	// Check for invalid keys
	for key := range req {
		if !JSONRPCRequestKeys[key] {
			return fmt.Errorf("invalid key: %s", key)
		}
	}

	return nil
}

// ValidateJSONRPCResponse validates a JSON-RPC 2.0 response
func ValidateJSONRPCResponse(response map[string]interface{}) error {
	// Must have either result or error, but not both
	hasResult := false
	hasError := false

	if _, ok := response["result"]; ok {
		hasResult = true
	}
	if _, ok := response["error"]; ok {
		hasError = true
	}

	if !hasResult && !hasError {
		return fmt.Errorf("response must have either 'result' or 'error'")
	}

	if hasResult && hasError {
		return fmt.Errorf("response cannot have both 'result' and 'error'")
	}

	// Check jsonrpc version
	if jsonrpc, ok := response["jsonrpc"].(string); ok {
		if jsonrpc != "2.0" {
			return fmt.Errorf("jsonrpc must be '2.0'")
		}
	}

	return nil
}

// IsValidNonErrorResponse checks if response is valid and non-error
func IsValidNonErrorResponse(response map[string]interface{}) bool {
	if err := ValidateJSONRPCResponse(response); err != nil {
		return false
	}

	// Must have result (not error)
	_, hasResult := response["result"]
	_, hasError := response["error"]

	return hasResult && !hasError
}

func isValidIDType(v interface{}) bool {
	kind := reflect.TypeOf(v).Kind()
	for _, validKind := range IDTypes {
		if kind == validKind {
			return true
		}
	}
	return false
}

func isValidParamsType(v interface{}) bool {
	if v == nil {
		return true
	}
	kind := reflect.TypeOf(v).Kind()
	for _, validKind := range ParamsTypes {
		if kind == validKind {
			return true
		}
	}
	return false
}

// Broadcast transaction method names
var BroadcastTransactionMethods = map[string]bool{
	"broadcast_transaction":             true,
	"broadcast_transaction_synchronous": true,
}

// IsGetBlockRequest checks if the request is a get_block request
func IsGetBlockRequest(req *request.JSONRPCRequest) bool {
	if req == nil || req.URN == nil {
		return false
	}
	return (req.URN.Namespace == "steemd" || req.URN.Namespace == "appbase") &&
		req.URN.Method == "get_block"
}

// IsGetBlockHeaderRequest checks if the request is a get_block_header request
func IsGetBlockHeaderRequest(req *request.JSONRPCRequest) bool {
	if req == nil || req.URN == nil {
		return false
	}
	return (req.URN.Namespace == "steemd" || req.URN.Namespace == "appbase") &&
		req.URN.Method == "get_block_header"
}

// IsBroadcastTransactionRequest checks if the request is a broadcast transaction request
func IsBroadcastTransactionRequest(req *request.JSONRPCRequest) bool {
	if req == nil || req.URN == nil {
		return false
	}
	return BroadcastTransactionMethods[req.URN.Method]
}

// BlockNumFromID extracts block number from block ID (first 8 hex digits)
func BlockNumFromID(blockID string) (int, error) {
	if len(blockID) < 8 {
		return 0, fmt.Errorf("block ID too short: %s", blockID)
	}
	blockNumStr := blockID[:8]
	blockNum, err := strconv.ParseInt(blockNumStr, 16, 32)
	if err != nil {
		return 0, fmt.Errorf("failed to parse block number from ID: %w", err)
	}
	return int(blockNum), nil
}

// IsValidGetBlockResponse validates that the response matches the get_block request
func IsValidGetBlockResponse(req *request.JSONRPCRequest, response map[string]interface{}) bool {
	if !IsGetBlockRequest(req) {
		return false
	}

	if !IsValidNonErrorResponse(response) {
		return false
	}

	// Check if result is nil (block doesn't exist yet)
	result, ok := response["result"].(map[string]interface{})
	if !ok || result == nil {
		return false
	}

	// Extract block_id from response
	var blockID string
	if id, ok := result["block_id"].(string); ok {
		blockID = id
	} else if block, ok := result["block"].(map[string]interface{}); ok {
		if id, ok := block["block_id"].(string); ok {
			blockID = id
		}
	}

	if blockID == "" {
		return false
	}

	// Extract block number from request params
	var requestBlockNum interface{}
	switch params := req.URN.Params.(type) {
	case []interface{}:
		if len(params) > 0 {
			requestBlockNum = params[0]
		}
	case map[string]interface{}:
		requestBlockNum = params["block_num"]
	}

	if requestBlockNum == nil {
		return false
	}

	// Convert request block num to int
	var reqBlockNum int
	switch v := requestBlockNum.(type) {
	case int:
		reqBlockNum = v
	case float64:
		reqBlockNum = int(v)
	case string:
		num, err := strconv.Atoi(v)
		if err != nil {
			return false
		}
		reqBlockNum = num
	default:
		return false
	}

	// Extract block number from response block ID
	respBlockNum, err := BlockNumFromID(blockID)
	if err != nil {
		return false
	}

	return reqBlockNum == respBlockNum
}

// LimitCustomJSONOpLength validates custom JSON operation length
func LimitCustomJSONOpLength(ops []interface{}, sizeLimit int) error {
	for _, op := range ops {
		opList, ok := op.([]interface{})
		if !ok || len(opList) < 2 {
			continue
		}

		opType, ok := opList[0].(string)
		if !ok || opType != "custom_json" {
			continue
		}

		opData, ok := opList[1].(map[string]interface{})
		if !ok {
			continue
		}

		jsonStr, ok := opData["json"].(string)
		if !ok {
			continue
		}

		// Check UTF-8 encoded length
		jsonBytes := []byte(jsonStr)
		if len(jsonBytes) > sizeLimit {
			return errors.NewCustomJSONOpLengthError(len(jsonBytes), sizeLimit)
		}
	}

	return nil
}

// LimitCustomJSONAccount validates custom JSON account blacklist
func LimitCustomJSONAccount(ops []interface{}, blacklistAccounts map[string]bool) error {
	if blacklistAccounts == nil || len(blacklistAccounts) == 0 {
		return nil
	}

	for _, op := range ops {
		opList, ok := op.([]interface{})
		if !ok || len(opList) < 2 {
			continue
		}

		opType, ok := opList[0].(string)
		if !ok || opType != "custom_json" {
			continue
		}

		opData, ok := opList[1].(map[string]interface{})
		if !ok {
			continue
		}

		// Check required_posting_auths
		if postingAuths, ok := opData["required_posting_auths"].([]interface{}); ok {
			for _, auth := range postingAuths {
				if authStr, ok := auth.(string); ok {
					if blacklistAccounts[authStr] {
						return errors.NewLimitsError(fmt.Sprintf("account %s is blacklisted", authStr))
					}
				}
			}
		}
	}

	return nil
}

// LimitBroadcastTransactionRequest validates broadcast transaction request limits
func LimitBroadcastTransactionRequest(req *request.JSONRPCRequest, limits map[string]interface{}) error {
	if !IsBroadcastTransactionRequest(req) {
		return nil
	}

	// Extract operations from request params
	var operations []interface{}
	switch params := req.URN.Params.(type) {
	case []interface{}:
		if len(params) > 0 {
			if trx, ok := params[0].(map[string]interface{}); ok {
				if ops, ok := trx["operations"].([]interface{}); ok {
					operations = ops
				}
			}
		}
	case map[string]interface{}:
		if trx, ok := params["trx"].(map[string]interface{}); ok {
			if ops, ok := trx["operations"].([]interface{}); ok {
				operations = ops
			}
		}
	}

	if len(operations) == 0 {
		return nil
	}

	// Filter custom_json operations
	var customJSONOps []interface{}
	for _, op := range operations {
		opList, ok := op.([]interface{})
		if !ok || len(opList) < 2 {
			continue
		}
		if opType, ok := opList[0].(string); ok && opType == "custom_json" {
			customJSONOps = append(customJSONOps, op)
		}
	}

	if len(customJSONOps) == 0 {
		return nil
	}

	// Validate custom JSON op length (default limit: 8192 bytes)
	sizeLimit := 8192
	if limit, ok := limits["custom_json_size_limit"].(float64); ok {
		sizeLimit = int(limit)
	} else if limit, ok := limits["custom_json_size_limit"].(int); ok {
		sizeLimit = limit
	}

	if err := LimitCustomJSONOpLength(customJSONOps, sizeLimit); err != nil {
		return err
	}

	// Validate account blacklist
	blacklistAccounts := make(map[string]bool)
	if accounts, ok := limits["accounts_blacklist"].([]interface{}); ok {
		for _, acc := range accounts {
			if accStr, ok := acc.(string); ok {
				blacklistAccounts[accStr] = true
			}
		}
	} else if accounts, ok := limits["accounts_blacklist"].([]string); ok {
		for _, acc := range accounts {
			blacklistAccounts[acc] = true
		}
	}

	return LimitCustomJSONAccount(customJSONOps, blacklistAccounts)
}

